### PR TITLE
fix: flush after a single event is sent

### PR DESCRIPTION
**Pull request by DanielSidhion**
## Description of the change

When the first event is enqueued, a flush is triggered, but not awaited. If a caller calls flush after sending a single event, the call succeeds and the callback is called, even though there is a pending flush to be resolved.

This commit fixes this by chaining the callback received in the flush call to the flush already in progress.

**Note**: your pre-commit hook is currently failing linting on a lot of places unrelated to this PR. I checked that this PR doesn't introduce any new linting issues, but I also didn't fix all linting issues in this PR because it's out of its scope.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

No current related issues, but this might solve #79.

### Development

- [ ] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: index.js
Comment (MoumitaM) [inline index.js:611]:
```suggestion
          .then((data) => {
            callback();
            return Promise.resolve(data);
          });
```

File: index.js
Comment (MoumitaM) [inline index.js:612]: Add a catch block.

**Comment by MoumitaM** [comment]
Hi @DanielSidhion, Please sign this [CLA](https://rudderlabs.wufoo.com/forms/rudderlabs-contributor-license-agreement)

File: index.js
Comment (DanielSidhion) [reply index.js:611]: I don't think this should be done. When we get to this code, the queue is empty, which means the callback for this call should resolve with no data. The only reason we're making the callback wait for the pending flush is so clients who are relying on flush() returning to shut down can also wait for pending flushes.

I can make this change, but it would mean that two callbacks to flush() might contain the same data. This might break certain cases. Let me know what you think.

File: index.js
Comment (DanielSidhion) [reply index.js:612]: How do you expect errors to be handled in this case? Do we just propagate the error again to all other callbacks?
The pending promise already contains a catch block with all the proper error handling, which will call the "original" callback if needed.

This is the scenario I have in mind:
- There is a pending flush already, with a callback from whatever code called `flush()` to create the pending flush.
- Some other consumer of the lib wants to flush as well, but no new events were added to the queue (this is what triggers the code in this PR).
- If the pending flush fails and we propagate the error, both the first callback and the callback from the other consumer will get the error.

I'll defer to what you think is best for the library to do, but I'm leaning towards resolving the callback without an error (just a `callback()` call) and throw the error again to keep it going through the chain. This "makes sense" to me because the `flush()` call on an empty queue wasn't the one that had the error, so in that sense we're maintaining the assumption that `flush()` on an empty queue will just call `callback()` without ceremonies.

**Comment by DanielSidhion** [comment]
Hi @MoumitaM, apologies for the delay there, just had some time to look at this. CLA is signed, waiting for your feedback so I can push changes on this PR.

File: index.js
Comment (MoumitaM) [reply index.js:612]: Make sense.

File: index.js
Comment (MoumitaM) [reply index.js:611]: Let's do it without resolving data.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**
